Reduce Max Row Improvement

ncvetkovicTT opened the pull request:
### Ticket
#22870 

### Problem description
LLK reduce op on FPU has some improvement potential. SDPA needs reduce(MAX, ROW) to be optimized. This PR provides new APIs and LLK implementations for reduce(MAX, ROW), supporting regular tile formats, and both DEST formats (16-bit and 32-bit).

### What's changed
I have made new set of LLKs and APIs for reducing a tile across rows with MAX operation. The speedup is achieved by several small changes and the following major changes:
- Reduction is performed on a whole tile stored in SrcA rather than the previous face-by-face method.
- Unpacker moves a whole tile in a single instruction, rather than moving face by face.
- A block variant is introduced, where Unpacker and Math utilize MOPs and Replay buffers to reduce across several tile columns of the tile block, i.e. block_ct_dim is provided to the LLKs to construct MOPs.
- A single face of the scaler is unpacked into SrcB and is reused for the duration of reduction across a whole tile block row.
- Test called test_sdpa_reduce_c tests the usage of the API and is added to pipeline.

LLK PR: https://github.com/tenstorrent/tt-llk/pull/724

### Checklist
- [x] [All post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml) CI passes - https://github.com/tenstorrent/tt-metal/actions/runs/19147194235
- [x] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml) CI with demo tests passes (if applicable) - https://github.com/tenstorrent/tt-metal/actions/runs/19140200307
- [ ] [Model regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml) CI passes (if applicable) - N/A
- [ ] [Device performance regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml) CI passes (if applicable) -N/A
- [ ] (For models and ops writers) [Single-card demo tests](https://github.com/tenstorrent/tt-metal/actions/workflows/single-card-demo-tests.yaml) CI passes (if applicable) See [recommended dev flow](https://github.com/tenstorrent/tt-metal/blob/main/models/docs/MODEL_ADD.md#a-recommended-dev-flow-on-github-for-adding-new-models). - N/A
- [ ] [Galaxy quick](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-quick.yaml) CI passes (if applicable)
- [ ] [Galaxy demo tests, for Llama](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-demo-tests.yaml) CI passes, if applicable, because of current Llama work
- [ ] (For runtime and ops writers) [T3000 unit tests](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-unit-tests.yaml) CI passes (if applicable, since this is run on push to main)
- [ ] (For models and ops writers) [T3000 demo tests](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-demo-tests.yaml) CI passes (if applicable, since this is required for release)
- [x] New/Existing tests provide coverage for changes

cglagovichTT committed:
#22870: Separate reduce row max API
- Create performance unit tests and add kernel profile scopes
- Same Tensix instructions just separate, works

ncvetkovicTT committed:
#22870: Hoist reduce_uninit out

ncvetkovicTT committed:
#22870: Add a block variant

ncvetkovicTT commented:
@cglagovichTT Sorry, it had many build errors, refrain from integrating until I push build fix. So sorry.

ncvetkovicTT committed:
#22870: Revert changes that would break CI

cglagovichTT committed:
Modify SDPA reduce_max unit test to use reduce_block_max_row LLK

ncvetkovicTT committed:
#22870: Add support for 32-bit DEST

ncvetkovicTT committed:
#0: Revert changes

ncvetkovicTT committed:
#22870: Add test for reduce_c and support for BH

ncvetkovicTT committed:
#0: Clang Tidy

fvranicTT commented on tests/tt_metal/tt_metal/test_kernels/misc/sdpa/reduce_c/compute.cpp:
```diff
@@ -0,0 +1,39 @@
+    constexpr uint32_t Sk_chunk_t = get_compile_time_arg_val(5);
+    constexpr uint32_t do_eltwise = get_compile_time_arg_val(6);
+
+    // Init compute?
```
? the ? 😀

cglagovichTT commented on tests/tt_metal/tt_metal/CMakeLists.txt:
```diff
@@ -36,6 +36,7 @@ set(TT_METAL_TESTS_SRCS
     test_core_range_set.cpp
     test_compile_sets_kernel_binaries.cpp
     test_compile_program.cpp
+    test_sdpa_reduce_c.cpp
```
Do you know if all tests under `tests/tt_metal/tt_metal` are picked up by some pipeline or do we need to add this new test specifically to a pipeline?

fvranicTT replied:
They are not, you need to explicitly add your test [here](https://github.com/tenstorrent/tt-metal/blob/main/tests/scripts/run_tt_metal.py).

ncvetkovicTT committed:
#22870: Add test_sdpa_reduce_c to pipeline

ncvetkovicTT committed:
#22870: Remove slow dispatch API from test_sdpa_reduce_c.cpp

ncvetkovicTT commented on tests/tt_metal/tt_metal/test_sdpa_reduce_c.cpp:
```diff
@@ -0,0 +1,348 @@
+        tt_metal::detail::WriteToBuffer(identity_scale_buffer, identity_scale_uint_vec);
+
+        // Execute
+        tt_metal::detail::LaunchProgram(device, program, true);
```
Thanks! Could you take a look again?

fvranicTT commented on tests/tt_metal/tt_metal/test_sdpa_reduce_c.cpp:
```diff
@@ -0,0 +1,356 @@
+        }
+    }
+
+    mesh_device.reset();
```
nit: no need for this, mesh_device is a shared pointer that will get deallocated automatically once it goes out of scope.

ncvetkovicTT commented on tests/scripts/run_tt_metal.py:
```diff
@@ -91,6 +91,7 @@
     TestEntry("tt_metal/tests/test_multi_core_kernel", "test_multi_core_kernel"),
     TestEntry("tt_metal/tests/test_interleaved_layouts", "test_interleaved_layouts"),
     TestEntry("tt_metal/tests/test_interleaved_l1_buffer", "test_interleaved_l1_buffer"),
+    TestEntry("tt_metal/tests/test_sdpa_reduce_c", "test_sdpa_reduce_c"),
```
Idk but it works when it's in fast-dispatch cat. @jbaumanTT any concerns with moving this line to TT_METAL_FAST_DISPATCH_TEST_ENTRIES?

ncvetkovicTT committed:
#22870: Add specialized reduce_row_max unpacking operations for Blackhole and Wormhole B0

ncvetkovicTT committed:
#22870: Transpose at the end of block variant

nvelickovicTT commented on tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_unpack_AB_api.h:
```diff
@@ -114,3 +134,83 @@ inline void llk_unpack_AB_reduce_init(
+    std::uint32_t base_address_b = get_local_cb_interface(operandB_id).fifo_rd_ptr - 1;
+
+    // Always tile index 0 for operandB
+    // TTI_UNPACR(SrcB, 0b00000000 /* Z_ch0_inc and Z_ch1_inc */, 0, 0, 0, 1, 1 /* Set Dvalid */,
```
Remove dead code.

ncvetkovicTT committed:
#22870: Move unpack init from LLAPI to LLK

ncvetkovicTT committed:
#22870: Add documentation to specialized reduce functions:
- Remove SDPA references
- Explain ADDR MODs
- Mention prerequisites and assumptions

ncvetkovicTT committed:
#22870: Move specialized reduce functions to separate files
- Add tests for reduce_max_row and reduce_block_max_row

ncvetkovicTT committed:
#22870: Remove single-tile variant

ncvetkovicTT committed:
#0: UMD revert

ncvetkovicTT committed:
#22870: Fix copyright headers

fvranicTT commented on tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_math_reduce_custom_api.h:
```diff
@@ -0,0 +1,52 @@
+// SPDX-FileCopyrightText: © 2025 Tenstorrent Inc.
```
```suggestion
// SPDX-FileCopyrightText: © 2025 Tenstorrent AI ULC
```
And so on :)

fvranicTT replied:
This should be automated anyway. No point in wasting brain cells on something like this. In some repos it is automated 😉

fvranicTT commented on tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_unpack_AB_reduce_custom_api.h:
So custom will become a literal 🕳️ for all things reduce that are not classic? 😁

ncvetkovicTT committed:
#0: Copyright updates
